proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,558 @@
+
+If the *primary_no_array_creation_expression* of an *element_access* is a value of an *array_type*, the *element_access* is an array access ([§11.7.10.2](expressions.md#117102-array-access)). **If the *primary_no_array_creation_expression* of an *element_access* is a variable or value of a fixed-size buffer type and the *argument_list* consists of a single argument, the *element_access* is a fixed-size buffer element access.** Otherwise, the *primary_no_array_creation_expression* shall be a variable or value of a class, struct, or interface type that has one or more indexer members, in which case the *element_access* is an indexer access ([§11.7.10.3](expressions.md#117103-indexer-access)).
+
+#### Fixed-size buffer element access
+
+For a fixed-size buffer element access, the *primary_no_array_creation_expression* of the *element_access* must be a variable or value of a fixed-size buffer type. Furthermore, the *argument_list* of a fixed-size buffer element access is not allowed to contain named arguments. The *argument_list* must contain a single expression, and the expression must be 
+- of type `int`, or
+- implicitly convertible to `int`, or
+- implicitly convertible to ```System.Index```, or 
+- implicitly convertible to ```System.Range```. 
+
+##### When the expresion type is int
+
+If *primary_no_array_creation_expression* is a writable variable, the result of evaluating a fixed-size buffer element access is a writable variable
+equaivalent to invoking [`public ref T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.span-1.item?view=net-8.0) with
+that integer value on an instance of ```System.Span<T>``` returned by ```[UnscopedRef] public System.Span<T> AsSpan()``` method on *primary_no_array_creation_expression*. 
+
+If *primary_no_array_creation_expression* is a readonly variable, the result of evaluating a fixed-size buffer element access is a readonly variable
+equaivalent to invoking [`public ref readonly T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.item?view=net-8.0) with
+that integer value on an instance of ```System.ReadOnlySpan<T>``` returned by ```[UnscopedRef] public readonly System.ReadOnlySpan<T> AsReadOnlySpan()```
+method on *primary_no_array_creation_expression*. 
+
+If *primary_no_array_creation_expression* is a value, the result of evaluating a fixed-size buffer element access is a value
+equaivalent to invoking [`public ref readonly T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.item?view=net-8.0) with
+that integer value on an instance of ```System.ReadOnlySpan<T>``` returned by ```[UnscopedRef] public readonly System.ReadOnlySpan<T> AsReadOnlySpan()```
+method on *primary_no_array_creation_expression*. 
+
+For example:
+``` C#
+void M1(Buffer10<int> x)
+{
+    ref int a = ref x[0]; // Ok, equivalent to `ref int a = ref x.AsSpan()[0]`
```

Comment:
The JIT should already be able to turn this into the equivalent of `ref x._element0`, but we should double-check it indeed does always do all the necessary inlining and the like.